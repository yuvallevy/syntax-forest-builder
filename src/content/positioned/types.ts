import { Id, IdMap, NodeCommon, StringSlice, TreeCommon } from '../types';

type PlotX = number;
type PlotY = number;
type TreeX = number;
type TreeY = number;
type Width = number;

export type PlotCoords = {
  plotX: PlotX;
  plotY: PlotY;
};

export type PositionInTree = {
  treeX: TreeX;
  treeY: TreeY;
};

type WithPositionInTree = { position: PositionInTree };

export type PositionedBranchingNode = NodeCommon & WithPositionInTree & {
  children: Id[];
};

export type PositionedTerminalNode = NodeCommon & WithPositionInTree & {
  slice: StringSlice;
  triangle?: {
    treeX1: TreeX;
    treeX2: TreeX;
  }
};

export type PositionedStrandedNode = NodeCommon & WithPositionInTree;

export type PositionedNode = PositionedBranchingNode | PositionedTerminalNode | PositionedStrandedNode;

export type PositionedTree = TreeCommon & {
  nodes: IdMap<PositionedNode>;
  position: PlotCoords;
  width: Width;
};

export type PositionedPlot = { trees: IdMap<PositionedTree> };
