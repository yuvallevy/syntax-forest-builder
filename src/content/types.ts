export type Id = string;
export type Sentence = string;
export type NodeLabel = string;

type SliceStart = number;
type SliceEndExclusive = number;

export type StringSlice = [SliceStart, SliceEndExclusive];

export type IdMap<T> = Record<Id, T>;

export type NodeIndicatorInPlot = {
  treeId: Id;
  nodeId: Id;
};

export type TreeCommon = {
  sentence: Sentence;
}

export type NodeCommon = {
  label: NodeLabel;
};
