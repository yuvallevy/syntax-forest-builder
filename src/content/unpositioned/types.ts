import { Id, IdMap, NodeCommon, StringSlice, TreeCommon } from '../types';

type DPlotX = number;
type DPlotY = number;
type DTreeX = number;
type DTreeY = number;

export type PlotCoordsOffset = {
  dPlotX: DPlotX;
  dPlotY: DPlotY;
};

export type UnpositionedPlot = { trees: IdMap<UnpositionedTree>; };

export type UnpositionedTree = TreeCommon & {
  nodes: IdMap<UnpositionedNode>;
  offset: PlotCoordsOffset;
};

type WithOffsetInTree = {
  offset: {
    dTreeX: DTreeX;
    dTreeY: DTreeY;
  };
};

export type UnpositionedBranchingNode = NodeCommon & WithOffsetInTree & {
  children: Id[];
};

export type UnpositionedTerminalNode = NodeCommon & WithOffsetInTree & {
  slice: StringSlice;
  triangle: boolean;
};

export type UnpositionedStrandedNode = NodeCommon & WithOffsetInTree & {
  formerDescendants?: IdMap<UnpositionedNode>;
  formerSlice?: StringSlice;
  formerlyTriangle?: boolean;
};

export type UnpositionedNode = UnpositionedBranchingNode | UnpositionedTerminalNode | UnpositionedStrandedNode;

export const isBranching = (node: UnpositionedNode): node is UnpositionedBranchingNode =>
  'children' in node;

export const isTerminal = (node: UnpositionedNode): node is UnpositionedTerminalNode =>
  'slice' in node;
