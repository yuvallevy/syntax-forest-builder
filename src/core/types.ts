export type Id = string;
export type Sentence = string;
export type NodeLabel = string;

type SliceStart = number;
type SliceEndExclusive = number;
type PlotX = number;
type PlotY = number;
type TreeX = number;
type TreeY = number;
type DPlotX = number;
type DPlotY = number;
type DTreeX = number;
type DTreeY = number;
type Width = number;

export type NodeSlice = [SliceStart, SliceEndExclusive];

export type IdMap<T> = Record<Id, T>;

export type UnpositionedForest = { plots: IdMap<UnpositionedPlot>; };

export type UnpositionedPlot = { trees: IdMap<UnpositionedTree>; };

type TreeCommon = {
  sentence: Sentence;
}

export type UnpositionedTree = TreeCommon & {
  nodes: IdMap<UnpositionedNode>;
  offset: { dPlotX: DPlotX; dPlotY: DPlotY; };
};

export type NodeCommon = {
  label: NodeLabel;
};

type WithOffsetInTree = {
  offset: {
    dTreeX: DTreeX;
    dTreeY: DTreeY;
  };
};

export type UnpositionedBranchingNode = NodeCommon & WithOffsetInTree & {
  children: IdMap<UnpositionedNode>;
};

export type UnpositionedTerminalNode = NodeCommon & WithOffsetInTree & {
  slice: NodeSlice;
  triangle: boolean;
};

export type UnpositionedNode = UnpositionedBranchingNode | UnpositionedTerminalNode;

export const isBranching = (node: UnpositionedNode): node is UnpositionedBranchingNode =>
  'children' in node;

export type PositionInTree = {
  treeX: TreeX;
  treeY: TreeY;
};

type WithPositionInTree = { position: PositionInTree };

export type PositionedBranchingNode = NodeCommon & WithPositionInTree & {
  children: IdMap<PositionedNode>;
};

export type PositionedTerminalNode = NodeCommon & WithPositionInTree & {
  triangle?: {
    treeX1: TreeX;
    treeX2: TreeX;
  }
};

export type PositionedNode = PositionedBranchingNode | PositionedTerminalNode;

export type PositionedTree = TreeCommon & {
  nodes: IdMap<PositionedNode>;
  position: {
    plotX: PlotX;
    plotY: PlotY;
  };
  width: Width;
};

export type PositionedPlot = { trees: IdMap<PositionedTree> };
