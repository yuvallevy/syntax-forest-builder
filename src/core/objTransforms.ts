export const transformValues = <K extends string, V, R>(record: Record<K, V>, valueTransform: (value: V) => R): Record<K, R> =>
  (Object.entries(record) as [K, V][]).reduce((transformedRecord, [key, value]) => ({
    ...transformedRecord,
    [key]: valueTransform(value),
  }), {} as Record<K, R>);

export const mapValues = <K extends string, V, R>(record: Record<K, V>, valueTransform: (value: V) => R): R[] =>
  (Object.values(record) as V[]).map(valueTransform);

export const mapEntries = <K extends string, V, R>(record: Record<K, V>, entryTransform: (entry: [K, V]) => R): R[] =>
  (Object.entries(record) as [K, V][]).map(entryTransform);

export const without = <T>(array: T[], elementToRemove: T): T[] =>
  array.reduce((accum, element) => element === elementToRemove ? accum : [...accum, element], [] as T[]);

export const union = <T>(array1: T[], array2: T[]): T[] =>
  array2.reduce((accum, element) => array1.includes(element) ? accum : [...accum, element], array1);

export const omitKey = <K extends string, V>(record: Record<K, V>, keyToOmit: K): Record<K, V> =>
  record.hasOwnProperty(keyToOmit)
    ? (Object.entries(record) as [K, V][]).reduce((transformedRecord, [key, value]) => key === keyToOmit ? transformedRecord : {
      ...transformedRecord,
      [key]: value,
    }, {} as Record<K, V>)
    : record;

export const isEmpty = (objOrArray: any): boolean => (objOrArray instanceof Array ? objOrArray : Object.keys(objOrArray)).length === 0;
