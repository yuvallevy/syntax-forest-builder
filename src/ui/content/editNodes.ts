import {
  Id, StringSlice, Sentence, TreeAndNodeId
} from '../../content/types';
import { InsertedNode, transformAllNodesInTree } from '../../content/unpositioned/manipulation';
import { isTerminal, UnpositionedNode, UnpositionedTree } from '../../content/unpositioned/types';

export type NodeSelectionInPlot = { nodes: TreeAndNodeId[] };
export type SliceSelectionInPlot = { treeId: Id, slice: StringSlice };
export type SelectionInPlot = NodeSelectionInPlot | SliceSelectionInPlot;

const isWordChar = (char: string) => /['A-Za-z\u00c0-\u1fff]/.test(char);

/**
 * Returns the range of the word that the given position is within.
 */
const getWordRange = (sentence: Sentence, position: number): [start: number, end: number] => {
  let start = position;
  while (start > 0 && isWordChar(sentence[start - 1])) start--;

  let end = position;
  while (end < sentence.length && isWordChar(sentence[end])) end++;

  return [start, end];
};

/**
 * Returns the definition for a new node based on the current selection.
 * If the selection is a slice of length 0, expands it to span the word where the cursor currently is,
 * using the given sentence.
 */
export const newNodeFromSelection = (selection: SelectionInPlot, sentence: Sentence): InsertedNode => {
  // A slice of the sentence is selected
  if ('slice' in selection) {
    const [sliceStart, sliceEnd] = selection.slice;
    const selectedSlice = sliceStart !== sliceEnd
      ? selection.slice
      : getWordRange(sentence, sliceStart);
    return {
      targetSlice: selectedSlice,
      label: '',
    };
  }
  // One or more nodes are selected
  const selectedNodeIds = selection.nodes.map(({ nodeId }) => nodeId);
  return {
    targetChildIds: selectedNodeIds,
    label: '',
  };
};

/**
 * Shifts the slice associated with a node by the given number of characters,
 * based on the kind of selection made and the operation done on it.
 * Node slices will not expand when adding characters, but may contract when removing characters.
 */
const shiftNodeSliceAfterChange =
  (oldSelection: StringSlice, shiftBy: number) =>
  (node: UnpositionedNode): UnpositionedNode => {
    if (                               // If:
      !isTerminal(node) ||             // this is not a terminal node, or
      shiftBy === 0 ||                 // the sentence length did not change, or
      node.slice[1] < oldSelection[0]  // the selection was entirely after the node, then
    ) return node;                     // no change is necessary

    const [oldNodeSliceStart, oldNodeSliceEnd] = node.slice;
    const newNodeSlice: StringSlice = [oldNodeSliceStart, oldNodeSliceEnd];
    if (oldSelection[0] === oldSelection[1]) {  // No selection, just a cursor
      const oldCursorPos = oldSelection[0];
      if (oldCursorPos === oldNodeSliceStart) {  // Cursor was at the beginning of the slice
        newNodeSlice[1] += shiftBy;
        if (shiftBy > 0) {  // If adding, move the slice forward; if removing, contract it
          newNodeSlice[0] += shiftBy;
        }
      }
      if (oldCursorPos === oldNodeSliceEnd) {  // Cursor was at the end of the slice
        if (shiftBy < 0) {  // If removing, contract the slice
          newNodeSlice[1] += shiftBy;
        }
      }
      if (oldCursorPos < oldNodeSliceStart) {  // Cursor was before the slice
        newNodeSlice[0] += shiftBy;
        newNodeSlice[1] += shiftBy;
      }
      if (oldNodeSliceStart < oldCursorPos && oldCursorPos < oldNodeSliceEnd) {  // Cursor was in the middle of the slice
        newNodeSlice[1] += shiftBy;
      }
    } else {  // A selection spanning at least one character was made
      console.error('Not implemented yet');
    }
    return { ...node, slice: newNodeSlice };
  }

/**
 * Responds to a change in the sentence associated with the given tree,
 * adapting existing nodes to the change if any exist.
 * @param newSentence The new sentence to use for the tree.
 * @param oldSelection The selection before the change was made.
 *   This is used to determine how exactly node ranges should change.
 */
export const handleLocalSentenceChange =
  (newSentence: Sentence, oldSelection: StringSlice) =>
  (tree: UnpositionedTree): UnpositionedTree => transformAllNodesInTree(
    shiftNodeSliceAfterChange(oldSelection, newSentence.length - tree.sentence.length)
  )({
    ...tree,
    sentence: newSentence,
  });
