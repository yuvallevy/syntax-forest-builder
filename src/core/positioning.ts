import { mapValues, transformValues } from './objTransforms';
import { PositionedNode, PositionedPlot, PositionedTree, Sentence, UnpositionedNode, UnpositionedPlot, UnpositionedTree } from './types';

const DEFAULT_TERMINAL_NODE_Y = -2;
const DEFAULT_TRIANGLE_NODE_Y = -20;
const DEFAULT_NODE_LEVEL_DIFF = -40;

type StrWidthFunc = (str: string) => number;

/**
 * Returns a node and its descendants, if any, with positions.
 */
const applyNodePosition = (strWidthFunc: StrWidthFunc) => (sentence: Sentence) => (node: UnpositionedNode | PositionedNode): PositionedNode => {
  if ('position' in node) {
    // This node already has a position
    return node;
  }

  if ('slice' in node) {
    // This is a terminal node
    const [sliceStart, sliceEnd] = node.slice;
    const widthBeforeSlice = strWidthFunc(sentence.slice(0, sliceStart));
    const sliceWidth = strWidthFunc(sentence.slice(sliceStart, sliceEnd));
    return {
      ...node,
      position: {
        treeX: widthBeforeSlice + (sliceWidth / 2) + node.offset.dTreeX,
        treeY: (node.triangle ? DEFAULT_TRIANGLE_NODE_Y : DEFAULT_TERMINAL_NODE_Y) + node.offset.dTreeY,
      },
      triangle: node.triangle ? {
        treeX1: widthBeforeSlice,
        treeX2: widthBeforeSlice + sliceWidth,
      } : undefined,
    };
  }

  // This is a branching node
  const positionedChildNodes = transformValues(node.children, applyNodePosition(strWidthFunc)(sentence));
  const childXs = mapValues(positionedChildNodes, node => node.position.treeX);
  const childYs = mapValues(positionedChildNodes, node => node.position.treeY);
  return {
    ...node,
    children: positionedChildNodes,
    position: {
      treeX: childXs.reduce((accum, cur) => accum + cur, 0) / childXs.length + node.offset.dTreeX,
      treeY: Math.min(...childYs) + DEFAULT_NODE_LEVEL_DIFF + node.offset.dTreeY,
    },
  };
}

/**
 * Returns a copy of the given tree with positions for all nodes.
 */
export const applyNodePositionsToTree = (strWidthFunc: StrWidthFunc) => (tree: UnpositionedTree): PositionedTree => ({
  ...tree,
  nodes: transformValues(tree.nodes, applyNodePosition(strWidthFunc)(tree.sentence)),
  position: {
    plotX: tree.offset.dPlotX,
    plotY: tree.offset.dPlotY,
  },
  width: strWidthFunc(tree.sentence),
});

/**
 * Returns a copy of the given plot with positions for all trees and nodes.
 */
export const applyNodePositionsToPlot = (strWidthFunc: StrWidthFunc) => (plot: UnpositionedPlot): PositionedPlot => ({
  ...plot,
  trees: transformValues(plot.trees, applyNodePositionsToTree(strWidthFunc)),
});
