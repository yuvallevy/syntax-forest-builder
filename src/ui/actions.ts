import { union } from '../core/objTransforms';
import { Id } from '../core/types';
import { UiState } from './state';

type UiAction =
  | { type: 'selectNodes', plotId: Id, treeIds: Id[], nodeIds: Id[], mode: 'set' | 'add' }
;

export const reducer = (state: UiState, action: UiAction): UiState => {
  switch (action.type) {
    case 'selectNodes':
      return {
        ...state,
        activePlotId: action.plotId,
        selectedTreeIds: action.mode === 'add' ? union(state.selectedTreeIds, action.treeIds) : action.treeIds,
        selectedNodeIds: action.mode === 'add' ? union(state.selectedNodeIds, action.nodeIds) : action.nodeIds,
      };
    default:
      return state;
  }
};
