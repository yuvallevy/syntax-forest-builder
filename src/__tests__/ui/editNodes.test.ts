import { describe, expect, it } from 'vitest';
import { StringSlice, Sentence, UnpositionedTree } from '../../core/types';
import { handleLocalSentenceChange, newNodeFromSelection } from '../../ui/editNodes';

describe('high-level tree editing routines', () => {
  describe('defining new nodes', () => {
    it('defines a new branching node', () => {
      expect(newNodeFromSelection({ nodes: [{ treeId: 'tree', nodeId: 'a' }, { treeId: 'tree', nodeId: 'b' }] }, ''))
        .toMatchObject({ targetChildIds: ['a', 'b'] });
    });

    describe('defines a new terminal node', () => {
      const testSentencesForWordRanges: [Sentence, number, StringSlice][] = [
        // The word in...                          ...at position...     ...is in range...
        ['She will buy the red sweater.',          2,                    [0, 3]],
        ['I have nowhere to go.',                  18,                   [18, 20]],
        ['That was a big surprise, wasn\'t it?',   27,                   [25, 31]],
        ['Look at the blackboard!',                13,                   [12, 22]],
        ['Look\u2003at\u2003the\u2003blackboard!', 13,                   [12, 22]],
      ];

      it.each(testSentencesForWordRanges)(
        'finds that the word in "%s" intersecting position %d is in range %p',
        (sentence, position, expectedSlice) => {
          expect(newNodeFromSelection({ treeId: 'test', slice: [position, position] }, sentence))
            .toMatchObject({ targetSlice: expectedSlice });
        });
    });
  });

  describe('accommodating sentence changes', () => {
    const tree: UnpositionedTree = {
      sentence: 'The dog jumped.',
      nodes: {
        'branch1': {
          label: 'NP',
          offset: { dTreeX: 0, dTreeY: 0 },
          children: ['term1', 'term2'],
        },
        'term1': {
          label: 'Det',
          offset: { dTreeX: 0, dTreeY: 0 },
          slice: [0, 3],
          triangle: false,
        },
        'term2': {
          label: 'N',
          offset: { dTreeX: 0, dTreeY: 0 },
          slice: [4, 7],
          triangle: false,
        },
        'term3': {
          label: 'VP',
          offset: { dTreeX: 0, dTreeY: 0 },
          slice: [8, 14],
          triangle: true,
        },
        'top': {
          label: 'S',
          offset: { dTreeX: 0, dTreeY: 5 },
          children: ['branch1', 'term3'],
        },
      },
      offset: { dPlotX: 0, dPlotY: 0 },
    };

    it('sets the sentence associated with a tree', () => {
      expect(handleLocalSentenceChange('The dogs jumped.', [7, 7])(tree).sentence).toBe('The dogs jumped.');
    });

    it.each([
      ['The dogs jumped.', [7, 7]],
      ['The wdog jumped.', [4, 4]],
      ['The doug jumped.', [6, 6]],
      ['The og jumped.', [5, 5]],  // Backward delete from index 5
      ['The og jumped.', [4, 4]],  // Forward delete from index 4
      ['The dg jumped.', [6, 6]],  // Backward delete from index 6
      ['The dg jumped.', [5, 5]],  // Forward delete from index 5
      ['The do jumped.', [7, 7]],  // Backward delete from index 7
      ['The do jumped.', [6, 6]],  // Forward delete from index 6
      ['The g jumped.', [6, 6]],  // Backward word delete from index 6
      ['The d jumped.', [5, 5]],  // Forward word delete from index 5
    ] as [Sentence, StringSlice][])('reassigns terminal node slices when changing sentence to "%s"', (newSentence, oldSelection) => {
      expect(handleLocalSentenceChange(newSentence, oldSelection)(tree).nodes).toMatchSnapshot();
    });
  });
});