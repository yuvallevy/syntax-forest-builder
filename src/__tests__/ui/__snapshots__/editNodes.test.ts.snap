// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`high-level tree editing routines accommodating sentence changes reassigns terminal node slices when changing sentence to "The d jumped." 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
      "term2",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "Det",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      3,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      4,
      5,
    ],
    "triangle": false,
  },
  "term3": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      6,
      12,
    ],
    "triangle": true,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term3",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`high-level tree editing routines accommodating sentence changes reassigns terminal node slices when changing sentence to "The dg jumped." 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
      "term2",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "Det",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      3,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      4,
      6,
    ],
    "triangle": false,
  },
  "term3": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      7,
      13,
    ],
    "triangle": true,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term3",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`high-level tree editing routines accommodating sentence changes reassigns terminal node slices when changing sentence to "The dg jumped." 2`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
      "term2",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "Det",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      3,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      4,
      6,
    ],
    "triangle": false,
  },
  "term3": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      7,
      13,
    ],
    "triangle": true,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term3",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`high-level tree editing routines accommodating sentence changes reassigns terminal node slices when changing sentence to "The do jumped." 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
      "term2",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "Det",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      3,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      4,
      6,
    ],
    "triangle": false,
  },
  "term3": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      7,
      13,
    ],
    "triangle": true,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term3",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`high-level tree editing routines accommodating sentence changes reassigns terminal node slices when changing sentence to "The do jumped." 2`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
      "term2",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "Det",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      3,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      4,
      6,
    ],
    "triangle": false,
  },
  "term3": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      7,
      13,
    ],
    "triangle": true,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term3",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`high-level tree editing routines accommodating sentence changes reassigns terminal node slices when changing sentence to "The dogs jumped." 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
      "term2",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "Det",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      3,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      4,
      7,
    ],
    "triangle": false,
  },
  "term3": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      9,
      15,
    ],
    "triangle": true,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term3",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`high-level tree editing routines accommodating sentence changes reassigns terminal node slices when changing sentence to "The doug jumped." 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
      "term2",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "Det",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      3,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      4,
      8,
    ],
    "triangle": false,
  },
  "term3": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      9,
      15,
    ],
    "triangle": true,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term3",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`high-level tree editing routines accommodating sentence changes reassigns terminal node slices when changing sentence to "The g jumped." 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
      "term2",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "Det",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      3,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      4,
      5,
    ],
    "triangle": false,
  },
  "term3": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      6,
      12,
    ],
    "triangle": true,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term3",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`high-level tree editing routines accommodating sentence changes reassigns terminal node slices when changing sentence to "The og jumped." 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
      "term2",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "Det",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      3,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      4,
      6,
    ],
    "triangle": false,
  },
  "term3": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      7,
      13,
    ],
    "triangle": true,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term3",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`high-level tree editing routines accommodating sentence changes reassigns terminal node slices when changing sentence to "The og jumped." 2`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
      "term2",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "Det",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      3,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      4,
      6,
    ],
    "triangle": false,
  },
  "term3": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      7,
      13,
    ],
    "triangle": true,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term3",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;

exports[`high-level tree editing routines accommodating sentence changes reassigns terminal node slices when changing sentence to "The wdog jumped." 1`] = `
Object {
  "branch1": Object {
    "children": Array [
      "term1",
      "term2",
    ],
    "label": "NP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
  },
  "term1": Object {
    "label": "Det",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      0,
      3,
    ],
    "triangle": false,
  },
  "term2": Object {
    "label": "N",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      5,
      8,
    ],
    "triangle": false,
  },
  "term3": Object {
    "label": "VP",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 0,
    },
    "slice": Array [
      9,
      15,
    ],
    "triangle": true,
  },
  "top": Object {
    "children": Array [
      "branch1",
      "term3",
    ],
    "label": "S",
    "offset": Object {
      "dTreeX": 0,
      "dTreeY": 5,
    },
  },
}
`;
